File: src/main/java/model/Node.java
package model;

import java.util.*;
import java.util.stream.Collectors;

public class Node {

    public static final String SEPARATOR_ARROW = " ->";
    public static final String SEPARATOR_EMPTY = "";

    private final List<Node> children;
    private final String nodeSymbol;

    public Node(String symbol) {
        this.children = new ArrayList<>();
        this.nodeSymbol = symbol;
    }

    public String getSymbol() {
        return nodeSymbol;
    }

    public void addChild(Node child) {
        children.add(child);
    }

    public int getCountOfChildren() {
        return children.size();
    }

    boolean isLeaf() {
        return getCountOfChildren() == 0;
    }

    private boolean isNode() {
        return !isLeaf();
    }

    public Node getChild(int number) {
        return children.get(number);
    }

    public Node getChild(int... numbers) {
        Node result = this;
        for (int i : numbers) {
            result = result.getChild(i);
        }
        return result;
    }

    public int getHeight() {
        if (isLeaf()) {
            return 0;
        }
        int max = 0;
        for (Node n : children) {
            if (max < n.getHeight()) {
                max = n.getHeight();
            }
        }
        return 1 + max;
    }

    /**
     * @param separator Trennezichen zwischen linker und rechter Seite der Produktionsregel
     * @return Produktionsregel für diesen Node
     */
    String getProductions(String separator) {
        StringBuilder builder = new StringBuilder();
        builder.append(getSymbol());
        builder.append(separator);
        for (Node child : children) {
            builder.append(" ");
            builder.append(child.getSymbol());
        }
        builder.append("\n");
        return builder.toString();
    }

    public Set<String> getProductionsForWholeTree() {
        Set<String> productions = new LinkedHashSet<>();
        productions.add(getProductions(SEPARATOR_ARROW));
        children.stream().filter(Node::isNode)
                .forEach(node -> productions.addAll(node.getProductionsForWholeTree()));
        return productions;
    }

    public List<String> getProductionsWithDuplicates() {
        List<String> productions = new LinkedList<>();
        productions.add(getProductions(SEPARATOR_EMPTY));
        List<Node> deviations = children.stream().filter(Node::deviatesToNonterminal).collect(Collectors.toList());
        deviations.forEach(node -> productions.addAll(node.getProductionsWithDuplicates()));
        return productions;
    }

    String yield() {
        if (isLeaf()) {
            return getSymbol() + " ";
        }
        StringBuilder builder = new StringBuilder();
        children.stream().map(Node::yield).forEach(builder::append);
        return builder.toString();
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();

        builder.append("(");
        builder.append(getSymbol());

        if (isNode()) {
            for (Node n : children) {
                builder.append(n.toString());
            }
        }

        builder.append(")");
        return builder.toString();
    }

    public boolean isTraceTree() {
        if (this.isLeaf()) return getSymbol().startsWith("*");
        return children.stream().allMatch(Node::isTraceTree);
    }

    public void removeTraceTrees() {
        children.stream().filter(n -> !n.isTraceTree()).forEach(Node::removeTraceTrees);
        List<Node> toRemove = children.stream().filter(Node::isTraceTree).collect(Collectors.toList());
        toRemove.forEach(children::remove);
    }

    boolean deviatesToTerminal() {
        return getCountOfChildren() == 1 && getChild(0).isLeaf();
    }

    private boolean deviatesToNonterminal() {
        return !deviatesToTerminal() && isNode();
    }

    public void writeToLexicon(WordUsageCounter counter) {
        children.forEach(n -> {
            if (n.isLeaf()) {
                counter.addWord(n.getSymbol(), this.getSymbol());
                return;
            }
            n.writeToLexicon(counter);
        });
    }

}

File: src/main/kotlin/model/Node.kt
package model

import java.util.*

class Node(val symbol: String) {

    private val children: MutableList<Node>

    init {
        this.children = ArrayList<Node>()
    }

    fun addChild(child: Node) {
        children.add(child)
    }

    val countOfChildren: Int
        get() = children.size

    val isLeaf: Boolean
        get() = countOfChildren == 0

    private val isNode: Boolean
        get() = !isLeaf

    fun getChild(number: Int): Node {
        return children[number]
    }

    fun getChild(vararg numbers: Int): Node {
        var result = this
        numbers.forEach { result = result.getChild(it) }
        return result
    }

    val height: Int
        get() = (children.maxBy { it.height }?.height ?: -1) + 1

    /**
     * @param separator Trennezichen zwischen linker und rechter Seite der Produktionsregel
     * *
     * @return Produktionsregel für diesen Node
     */
    fun getProductions(separator: String): String {
        val builder = StringBuilder()
        builder.append(symbol)
        builder.append(separator)
        children.forEach { builder.append(" ${it.symbol}") }
        return builder.toString()
    }

    fun getRegularProductions(separator: String): RegularRule {
        val builder = StringBuilder()
        builder.append(" $baseCategory(")
        val joiner = StringJoiner(" ")
        children.map { it.formatAsRegularSymbol() }
                .forEach { joiner.add(it) }
        builder.append("$joiner)")
        return RegularRule(symbol, separator, builder.toString())
    }

    fun formatAsRegularSymbol(): String {
        if (isLeaf) return symbol
        return "q_$symbol"
    }

    val baseCategory: String
        get() {
            val index = symbol.indexOfFirst { it == '-' || it == '=' }
            if (index > -1) {
                return symbol.substring(0, index)
            }
            return symbol
        }

    val productionsForWholeTree: MutableSet<String>
        get() {
            val productions = LinkedHashSet<String>()
            productions.add(getProductions(SEPARATOR_ARROW))
            children.filter { it.isNode }
                    .forEach { productions.addAll(it.productionsForWholeTree) }
            return productions
        }

    val regularProductionsForWholeTree: MutableSet<RegularRule>
        get() {
            val productions = LinkedHashSet<RegularRule>()
            productions.add(getRegularProductions(SEPARATOR_ARROW))
            children.filter { it.isNode }
                    .forEach { productions.addAll(it.regularProductionsForWholeTree) }
            return productions
        }

    val productionsWithDuplicates: MutableList<String>
        get() {
            val productions = LinkedList<String>()
            productions.add(getProductions(SEPARATOR_EMPTY))
            val deviations = LinkedHashSet<Node>()
            children.filter { it.deviatesToNonterminal() }
                    .toCollection(deviations)
            deviations.forEach { productions.addAll(it.productionsWithDuplicates) }
            return productions
        }

    fun yieldOperation(): String {
        if (isLeaf) {
            return symbol + " "
        }
        val builder = StringBuilder()
        children.map { it.yieldOperation() }.forEach { builder.append(it) }
        return builder.toString()
    }

    override fun toString(): String {
        val builder = StringBuilder()

        builder.append("(")
        builder.append(symbol)

        if (isNode) {
            children.forEach { builder.append(it.toString()) }
        }

        builder.append(")")
        return builder.toString()
    }

    val isTraceTree: Boolean
        get() {
            if (this.isLeaf) return symbol.startsWith("*")
            return children.all { it.isTraceTree }
        }

    fun removeTraceTrees() {
        children.filter { !it.isTraceTree }.forEach { it.removeTraceTrees() }
        val toRemove = LinkedList<Node>()
        children.filter { it.isTraceTree }.toCollection(toRemove)
        toRemove.forEach { children.remove(it) }
    }

    fun deviatesToTerminal(): Boolean {
        return countOfChildren == 1 && getChild(0).isLeaf
    }

    private fun deviatesToNonterminal(): Boolean {
        return !deviatesToTerminal() && isNode
    }

    fun writeToLexicon(counter: WordUsageCounter) {
        children.forEach {
            if (it.isLeaf) {
                counter.addWord(it.symbol, this.symbol)
                return
            }
            it.writeToLexicon(counter)
        }
    }

    companion object {
        val SEPARATOR_ARROW = " ->"
        val SEPARATOR_EMPTY = ""
    }

}
